flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -2424,11 +2438,55 @@ struct GlobalOpConversion : public FIROpConversion<fir::GlobalOp> {
     auto g = rewriter.create<mlir::LLVM::GlobalOp>(
         loc, tyAttr, isConst, linkage, global.sym_name(), initAttr);
     auto &gr = g.getInitializerRegion();
+    if (global.getType().isa<fir::SequenceType>() &&
+        !global.getType().dyn_cast<fir::SequenceType>().hasUnknownShape() &&
+        global.hasInitializationBody()) {
+      if (auto singleInsertOp = getSingleOpOfType<fir::InsertOnRangeOp>(
+              global.region().front())) {
```

Comment:
Do we want to consider the case of successive ranges?
```
a(1:10) = 0; a(11:20) = 1; ... a(91:100) = 9;
```

Reply:
I just did a quick test but it doesn't look like this is done in the global initializer region.

Reply:
Let me double check that.

Reply:
Here's an example:
```fortran
  type t
     integer i(500)
     integer j(500)
  end type t

  type(t), save :: lookForMe = t([(2, i=1,500)], [(52, j=1,500)])

  call x(lookForMe)  ! keep it in the output
end
```

That will generate this global
```mlir
fir.global internal @_QElookforme : !fir.type<_QTt{i:!fir.array<500xi32>,j:!fir.array<500xi32>}> {
  %c2_i32 = constant 2 : i32
  %c52_i32 = constant 52 : i32
  %0 = fir.undefined !fir.type<_QTt{i:!fir.array<500xi32>,j:!fir.array<500xi32>}>
  %1 = fir.undefined !fir.array<500xi32>
  %2 = fir.insert_on_range %1, %c2_i32, [0 : index, 499 : index] : (!fir.array<500xi32>, i32) -> !fir.array<500xi32>
  %3 = fir.insert_value %0, %2, ["i", !fir.type<_QTt{i:!fir.array<500xi32>,j:!fir.array<500xi32>}>] :(!fir.type<_QTt{i:!fir.array<500xi32>,j:!fir.array<500xi32>}>, !fir.array<500xi32>) -> !fir.type<_QTt{i:!fir.array<500xi32>,j:!fir.array<500xi32>}>
  %4 = fir.insert_on_range %1, %c52_i32, [0 : index, 499 : index] : (!fir.array<500xi32>, i32) -> !fir.array<500xi32>
  %5 = fir.insert_value %3, %4, ["j", !fir.type<_QTt{i:!fir.array<500xi32>,j:!fir.array<500xi32>}>] : (!fir.type<_QTt{i:!fir.array<500xi32>,j:!fir.array<500xi32>}>, !fir.array<500xi32>) -> !fir.type<_QTt{i:!fir.array<500xi32>,j:!fir.array<500xi32>}>
  fir.has_value %5 : !fir.type<_QTt{i:!fir.array<500xi32>,j:!fir.array<500xi32>}>
}
```

Reply:
Ok I didn't think of this use case. That could be converted to something like:

```
llvm.mlir.global internal @_QElookforme() : !llvm.struct<"_QTt", (array<500 x i32>, array<500 x i32>)> {
  %2 = llvm.mlir.undef : !llvm.struct<"_QTt", (array<500 x i32>, array<500 x i32>)>
  %3 = llvm.mlir.undef : !llvm.array<500 x i32>
  %4 = llvm.mlir.constant(dense<2> : vector<500xi32>) : !llvm.array<500 x i32>
  %5 = llvm.insertvalue %4, %2[0 : i32] : !llvm.struct<"_QTt", (array<500 x i32>, array<500 x i32>)>
  %6 = llvm.mlir.constant(dense<52> : vector<500xi32>) : !llvm.array<500 x i32>
  %7 = llvm.insertvalue %6, %5[1 : i32] : !llvm.struct<"_QTt", (array<500 x i32>, array<500 x i32>)>
  llvm.return %7 : !llvm.struct<"_QTt", (array<500 x i32>, array<500 x i32>)>
}
```

Reply:
@schweitzpgi I change a bit the code and it now works for more use cases out of the box. I just replace insert_on_range when the range covers fully the one from the attached type.